runtime/neumann/src/fee_handler.rs

```diff
@@ -0,0 +1,277 @@
+		Opposite = C::NegativeImbalance,
+	>,
+	C::NegativeImbalance: Imbalance<
+		<C as Currency<<T as frame_system::Config>::AccountId>>::Balance,
+		Opposite = C::PositiveImbalance,
+	>,
+	MC: MultiCurrency<<T as frame_system::Config>::AccountId>,
+	MC::CurrencyId: From<TokenId>,
+	MC::Balance: From<C::Balance>,
+	OU: OnUnbalanced<NegativeImbalanceOf<C, T>>,
+	TA: Get<T::AccountId>,
+	FCP: CallParser<CallOf<T>>,
+{
+	type LiquidityInfo = Option<NegativeImbalanceOf<C, T>>;
+	type Balance = <C as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+	/// Withdraw the prxedicted fee from the transaction origin.
+	///
+	/// Note: The `fee` already includes the `tip`.
+	fn withdraw_fee(
+		who: &T::AccountId,
+		call: &T::Call,
+		_info: &DispatchInfoOf<T::Call>,
+		fee: Self::Balance,
+		tip: Self::Balance,
+	) -> Result<Self::LiquidityInfo, TransactionValidityError> {
+		if fee.is_zero() {
+			return Ok(None)
+		}
+
+		let call_information = FCP::fee_information(call.clone());
+
+		if call_information.token_id == NATIVE_TOKEN_ID {
+			NativeFeeProcessor::<T, C>::withdraw_fee(who, fee, tip)
+		} else {
+			ForeignFeeProcessor::<T, MC, TA>::withdraw_fee(who, fee.into(), call_information)?;
+			Ok(None)
+		}
+	}
+
+	/// Hand the fee and the tip over to the `[OnUnbalanced]` implementation.
+	/// Since the predicted fee might have been too high, parts of the fee may
+	/// be refunded.
+	///
+	/// Note: The `corrected_fee` already includes the `tip`.
+	fn correct_and_deposit_fee(
+		who: &T::AccountId,
+		_dispatch_info: &DispatchInfoOf<T::Call>,
+		_post_info: &PostDispatchInfoOf<T::Call>,
+		corrected_fee: Self::Balance,
+		tip: Self::Balance,
+		already_withdrawn: Self::LiquidityInfo,
+	) -> Result<(), TransactionValidityError> {
+		if let Some(paid) = already_withdrawn {
+			// Calculate how much refund we should return
+			let refund_amount = paid.peek().saturating_sub(corrected_fee);
+			// refund to the the account that paid the fees. If this fails, the
+			// account might have dropped below the existential balance. In
+			// that case we don't refund anything.
+			let refund_imbalance = C::deposit_into_existing(who, refund_amount)
+				.unwrap_or_else(|_| C::PositiveImbalance::zero());
+			// merge the imbalance caused by paying the fees and refunding parts of it again.
+			let adjusted_paid = paid
+				.offset(refund_imbalance)
+				.same()
+				.map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+			// Call someone else to handle the imbalance (fee and tip separately)
+			let (tip, fee) = adjusted_paid.split(tip);
+			OU::on_unbalanceds(Some(fee).into_iter().chain(Some(tip)));
+		}
+		Ok(())
+	}
+}
+
+pub trait FeeConversion {
+	fn convert_fee_into_foreign(&self, fee: Balance) -> Option<Balance>;
+}
+impl FeeConversion for orml_asset_registry::AssetMetadata<Balance, CustomMetadata> {
+	fn convert_fee_into_foreign(&self, fee: Balance) -> Option<Balance> {
```

Comment:
Got you a failing test-
```
	#[test]
	fn what_happens_with_large_decimal_diff() {
		let asset = test_asset!(20, 10000:1);
		assert_eq!(asset.convert_fee_into_foreign(1000000000), Some(1000))
	}
```
integer math gets you in trouble when the decimal diff is large. I'm not sure how likely this is but I don't think the example I picked is totally crazy: token with 20 decimals that has a 10000:1 exchange rate with us.